nx/lib/nx.ex

```diff
@@ -6408,17 +6408,219 @@ defmodule Nx do
+            [15.0, 18.0, 21.0, 24.0]
+          ],
+          [
+            [30.0, 36.0, 42.0, 48.0],
+            [45.0, 54.0, 63.0, 72.0]
+          ]
+        ]
+      >
+
+    ### Dot product between a non-batched tensor and a batched tensor
+
+      iex> u = Nx.tensor([[1.0, 0.0, 1.0], [0.0, -1.0, 1.0]])
+      iex> v = Nx.tensor([[[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]], [[1, 1, 1, 0], [0, 1, 1, 1], [0, 0, 1, 1]]])
+      iex> Nx.dot(u, [-1], [], v, [1], [0])
+      #Nx.Tensor<
+        f32[2][2][4]
+        [
+          [
+            [2.0, 2.0, 2.0, 2.0],
+            [0.0, 0.0, 0.0, 0.0]
+          ],
+          [
+            [1.0, 1.0, 2.0, 1.0],
+            [0.0, -1.0, 0.0, 0.0]
+          ]
+        ]
+      >
+
+  """
+  @doc type: :ndim
+  def dot(t1, contract_axes1, batch_axes1, t2, contract_axes2, batch_axes2) do
     output_type = binary_type(t1, t2)
     %T{shape: s1, names: names1} = t1 = to_tensor(t1)
     %T{shape: s2, names: names2} = t2 = to_tensor(t2)
-    axes1 = Nx.Shape.normalize_axes(s1, axes1, names1)
-    axes2 = Nx.Shape.normalize_axes(s2, axes2, names2)
-    {output_shape, output_names} = Nx.Shape.zip_reduce(s1, axes1, names1, s2, axes2, names2)
+
+    # Axes normalization
+    c1 = Nx.Shape.normalize_axes(s1, contract_axes1, names1)
+    c2 = Nx.Shape.normalize_axes(s2, contract_axes2, names2)
+    b1 = Nx.Shape.normalize_axes(s1, batch_axes1, names1)
+    b2 = Nx.Shape.normalize_axes(s2, batch_axes2, names2)
+
+    left_batched? = b1 != []
+    right_batched? = b2 != []
+
+    # ensure normalized batch axis of left is valid value
+    if left_batched? and b1 != [0] do
+      msg = bad_batch_axis_message("left", batch_axes1, b1)
+      raise ArgumentError, msg
+    end
+
+    # ensure normalized batch axis of right is valid value
+    if right_batched? and b2 != [0] do
+      msg = bad_batch_axis_message("right", batch_axes2, b2)
+      raise ArgumentError, msg
+    end
+
+    # ensure batch dim sizes match if both tensors are batched
+    if left_batched? and right_batched? and elem(s1, 0) != elem(s2, 0) do
+      raise ArgumentError,
+            "dot batch dimension sizes must match, but the left " <>
+              "batch dimension of axes #{inspect(batch_axes1)} was [#{elem(s1, 0)}] " <>
+              "and the right batch dimension of axes #{inspect(batch_axes2)} was [#{elem(s2, 0)}]"
+    end
+
+    # ensure there is no conflict between left batch axes and left contract axes
+    if left_batched? and 0 in c1 do
+      raise ArgumentError,
+            batch_vs_contract_conflict_message("left", batch_axes1, b1, contract_axes1, c1)
+    end
+
+    # ensure there is no conflict between right batch axis and right contract axes
+    if right_batched? and 0 in c2 do
+      raise ArgumentError,
+            batch_vs_contract_conflict_message("right", batch_axes2, b2, contract_axes2, c2)
+    end
+
+    {output_shape, output_names} = dot_batch_output_shape(s1, c1, names1, b1, s2, c2, names2, b2)
```

Comment:
I think most of this should probably be combined into a dot shape rule. I've found it to be helpful for Axon to have some of the Nx shape rules available, and it might be necessary here as well